accord-core/src/main/java/accord/coordinate/Invalidate.java

```diff
@@ -22,126 +22,114 @@
-        this.preacceptTracker = new QuorumShardTracker(shard);
-    }
-
-    public static Invalidate invalidateIfNotWitnessed(Node node, TxnId txnId, RoutingKeys informKeys, RoutingKey invalidateWithKey, BiConsumer<Outcome, Throwable> callback)
-    {
-        return invalidate(node, txnId, informKeys, invalidateWithKey, PreAccepted, callback);
+        this.invalidateWithKeys = invalidateWithKeys;
+        Topologies topologies = node.topology().forEpoch(invalidateWithKeys, txnId.epoch);
+        this.tracker = new InvalidationTracker(topologies);
     }
 
-    public static Invalidate invalidate(Node node, TxnId txnId, RoutingKeys informKeys, RoutingKey invalidateWithKey, BiConsumer<Outcome, Throwable> callback)
+    public static Invalidate invalidate(Node node, TxnId txnId, RoutingKeys invalidateWithKeys, BiConsumer<Outcome, Throwable> callback)
     {
-        return invalidate(node, txnId, informKeys, invalidateWithKey, Accepted, callback);
+        Ballot ballot = new Ballot(node.uniqueNow());
+        Invalidate invalidate = new Invalidate(node, ballot, txnId, invalidateWithKeys, callback);
+        invalidate.start();
+        return invalidate;
     }
 
-    private static Invalidate invalidate(Node node, TxnId txnId, RoutingKeys informKeys, RoutingKey invalidateWithKey, Status recoverIfAtLeast, BiConsumer<Outcome, Throwable> callback)
+    private void start()
     {
-        Ballot ballot = new Ballot(node.uniqueNow());
-        Shard shard = node.topology().forEpochIfKnown(invalidateWithKey, txnId.epoch);
-        Invalidate invalidate = new Invalidate(node, shard, ballot, txnId, informKeys, invalidateWithKey, recoverIfAtLeast, callback);
-        node.send(shard.nodes, to -> new BeginInvalidation(txnId, invalidateWithKey, ballot), invalidate);
-        return invalidate;
+        node.send(tracker.nodes(), to -> new BeginInvalidation(to, tracker.topologies(), txnId, invalidateWithKeys, ballot), this);
     }
 
     @Override
     public synchronized void onSuccess(Id from, InvalidateReply reply)
     {
-        if (isDone || preacceptTracker.hasReachedQuorum())
+        if (isDone || isPrepareDone)
             return;
 
-        if (!reply.isOk())
+        replies.add(reply);
+        switch (tracker.recordSuccess(from, reply.isPromised(), reply.acceptedFastPath))
         {
-            InvalidateNack nack = (InvalidateNack) reply;
-            if (nack.homeKey != null)
-            {
-                node.ifLocalSince(contextFor(txnId), invalidateWithKey, txnId, safeStore -> {
-                    safeStore.command(txnId).updateHomeKey(safeStore, nack.homeKey);
-                }).addCallback(node.agent());
-            }
+            default: throw new AssertionError();
+            case Success:
+                invalidate();
+                break;
 
-            isDone = true;
-            callback.accept(null, new Preempted(txnId, null));
-            return;
-        }
+            case Failed:
```

Comment:
FYI this is still present. I think it really is worth mentioning that failure here can be something exceptional or it can be something like the transaction has made to much progress to be invalidated.

Reply:
I don't think it can be? Failed here means we literally had too many failures, or failed to get enough promises. I'll comment to this effect, anyway